### ISSUE TITLE
GCPPubSubQueue stopped receiving messages 

### ISSUE THREAD

**Issue by bradleyfalzon** [bradleyfalzon/gopherci#41]
At 2017/01/26 04:09 ACDT, we received an alert that the GCPPubSubQueue wasn't being processed. It was resolved at 2017/01/26 06:33 ACDT by restarting the gopherci process.

GopherCI itself is responsible for this, and was successfully running, receiving other webhooks, but a background goroutine failed to receive and process messages from the GCP PubSub queue.

1. main creates unbuffered channel
1. main passes a write only channel to GCPPubSubQueue, which is used to signal a job needs to run
1. main also receives a queue.Queuer from NewGCPPubSubQueue, which is provided to github package to add messages to the Google PubSub queue
1. main then starts a queueListen, which listens on the unbuffered channel for jobs and depending on the type, calls different methods to process it (blocking to ensure only 1 event is processed at a time).

The last successful job to be processed happened like so:

```
Jan 25 08:34:33 github: parsed webhook event: *github.PullRequestEvent
Jan 25 08:34:33 event: &github.PullRequestEvent{Action:(*string)(0xc420226410), Number:(*int)(0xc420226500), PullRequest:(*githu
b.PullRequest)(0xc4203d6360), Changes:(*github.EditChange)(nil), Repo:(*github.Repository)(0xc4201b4000), Sender:(*github.User)(0xc42052da20), Installation:(*github.Installation)(0xc4202743c0)}
Jan 25 08:34:33 prevent: &github.Installation{ID:(*int)(0xc42047e270), Account:(*github.User)(nil), AccessTokensURL:(*string)(ni
l), RepositoriesURL:(*string)(nil)}
Jan 25 08:34:33 GCPPubSubQueue: published a message with a message ID: 53103672313656
Jan 25 08:34:34 GCPPubSubQueue: processing ID 53103672313656, published at 2017-01-24 22:04:33.767 +0000 UTC
Jan 25 08:34:34 main: reading job type *github.PullRequestEvent
Jan 25 08:34:34 queue processing error: ignoring PR #16 action: "closed"
```

Note some columns have been removed that were redundant, duplicate date column, process IDs and hostnames

1. Firstly we received the webhook from GitHub: https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/internal/github/handlers.go#L32-L34
1. We then detected it was a *github.PullRequestEvent and therefore needed to be queued https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/internal/github/handlers.go#L39-L41
1. Queuer (in this case GCPPubSubQueue) successfully queued the message https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/internal/queue/gcp-pubsub.go#L110
1. Moments later a GCPPubSubQueue background goroutine picked up the job in the queue and added to the main queuer channel https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/internal/queue/gcp-pubsub.go#L138
1. main.queueListen received the message from the channel https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/main.go#L173
1. Detected the type and passed it along to the relevant handler, which detected the PR event was a close event and therefore nothing to do: https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/internal/github/handlers.go#L68-L70
1. main.queueListen received the error (note, this shouldn't be an error) https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/main.go#L182

That behaviour is expected, and at this point main.queueListen wouldn't blocked listening to the queueChan or context closing.

Over time, we continued to receive hook events from GitHub, but they were PushEvents (we don't currently support) or IntegrationInstallation events (they are processed immediately and without queues).

Next, we received:

```
Jan 26 04:03:53 github: parsed webhook event: *github.PullRequestEvent
Jan 26 04:03:53 event: &github.PullRequestEvent{Action:(*string)(0xc42047e790), Number:(*int)(0xc42047e890), PullRequest:(*github.PullRequest)(0xc4203d6240), Changes:(*github.EditChange)(nil), Repo:(*github.Repository)(0xc4201b4580), Sender:(*github.User)(0xc42052c160), Installation:(*github.Installation)(0xc42026b520)}
Jan 26 04:03:53 prevent: &github.Installation{ID:(*int)(0xc420477b90), Account:(*github.User)(nil), AccessTokensURL:(*string)(nil), RepositoriesURL:(*string)(nil)}
Jan 26 04:03:53 GCPPubSubQueue: published a message with a message ID: 53153322486157
```

Just like before:

1. We receive the hook from GitHub
1. GCPPubSubQueue publishes the message and receives an ID https://github.com/bradleyfalzon/gopherci/blob/06cdfb1a434976894c348fa4ebc2f089ca4ba094/internal/queue/gcp-pubsub.go#L110

But, unlike before, the background GCPPubSubQueue.listen goroutine never picks up the message.

Here's the logs with a grep for GCPPubSubQueue

```
Jan 25 08:34:33 GCPPubSubQueue: published a message with a message ID: 53103672313656
Jan 25 08:34:34 GCPPubSubQueue: processing ID 53103672313656, published at 2017-01-24 22:04:33.767 +0000 UTC
Jan 26 04:03:53 GCPPubSubQueue: published a message with a message ID: 53153322486157
Jan 26 04:16:35 GCPPubSubQueue: published a message with a message ID: 53153869748866
Jan 26 04:17:39 GCPPubSubQueue: published a message with a message ID: 53154023226121
Jan 26 04:26:56 GCPPubSubQueue: published a message with a message ID: 53154516087058
Jan 26 04:57:24 GCPPubSubQueue: published a message with a message ID: 53155951693825
``` 

We continue to receive hooks from GitHub, continue to add those to the queue, but we never receive them. GCPPubSubQueue.listen goroutine doesn't return without logging, and no other messages were received in the meantime.

The only other errors in between 'Jan 25 08:34:33' and 'Jan 26 04:03:53'

```
Jan 25 08:34:34 www1.c.gopherci-prod.internal gopherci[13275]: 2017/01/25 08:34:34 queue processing error: ignoring PR #16 action: "closed"
Jan 25 21:58:57 www1.c.gopherci-prod.internal gopherci[13275]: 2017/01/25 21:58:57 transport: http2Client.notifyError got notified that the client transport was broken EOF.
Jan 26 02:07:17 www1.c.gopherci-prod.internal gopherci[13275]: [mysql] 2017/01/26 02:07:17 packets.go:130: write tcp 10.138.0.2:44654->XXX.XXX.XXX.253:3306: write: broken pipe
Jan 26 02:07:17 www1.c.gopherci-prod.internal gopherci[13275]: [mysql] 2017/01/26 02:07:17 packets.go:130: write tcp 10.138.0.2:44654->XXX.XXX.XXX.253:3306: write: broken pipe
Jan 26 04:03:53 www1.c.gopherci-prod.internal gopherci[13275]: 2017/01/26 04:03:53 prevent: &github.Installation{ID:(*int)(0xc420477b90), Account:(*github.User)(nil), AccessTokensURL:(*string)(ni
l), RepositoriesURL:(*string)(nil)}
```

The packets.go errors just appears to be mysql driver related, the transport error maybe related as it appears to occur in https://github.com/grpc/grpc-go/blob/cb653e4b6150b81ba5157618b57c6f910a6a99f7/transport/http2_client.go#L1126

The grpc library is used by pubsub package: https://godoc.org/cloud.google.com/go/pubsub?imports

But we've received those messages quite a few times without an issue (count month day)

```
      1 Jan 23
     25 Jan 24
      9 Jan 25
```

So I'm really not sure what could have caused this. Monitoring does detect this issue (queue > n length for x period of time), and restarting the GopherCI process does also fix it.

I'll add some more debug to gopherci to determine whether the problem is internal, I'd like to add some kind of broadcast message to the queue continuously, so all works could log they received the message, this'd help us narrow down when the issue occurs.

Currently, we're only using a single instance for processing messages, and if this was just a client error, another server could process that job so customers aren't affected. We should do this anyway, but we didn't need it to handle the load just yet.

Note times are ACDT.

**Comment by bradleyfalzon** [comment]
This has occurred again, but due to the previous repository logging a large number of messages, most of the logs were suppressed.

**Comment by bradleyfalzon** [comment]
Happened again :/

```
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 github: pull request event: edited, installation id: 5912
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 GCPPubSubQueue: published a message with a message ID: 54562609476152
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 GCPPubSubQueue: processing ID 54562609476152, published at 2017-02-10 17:23:52.236 +0000 UTC
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 GCPPubSubQueue: ack'd ID 54562609476152
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 GCPPubSubQueue: adding ID 54562609476152 to job channel
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 GCPPubSubQueue: successfully added ID 54562609476152 to job queue
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 queueListen: reading job type *github.PullRequestEvent
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 pr action "edited" on repo https://github.com/owner/repo/pull/10
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 ignoring action "edited"
Feb 11 03:53:52 www1 gopherci[1205]: 2017/02/11 03:53:52 queueListen: finished processing in 173.954µs
Feb 11 04:07:02 www1 gopherci[1205]: 2017/02/11 04:07:02 github: ignored webhook event: *github.PushEvent
Feb 11 04:18:07 www1 gopherci[1205]: 2017/02/11 04:18:07 github: ignored webhook event: *github.PushEvent
Feb 12 14:56:03 www1 gopherci[1205]: 2017/02/12 14:56:03 github: ignored webhook event: *github.PushEvent
Feb 12 14:57:24 www1 gopherci[1205]: 2017/02/12 14:57:24 github: ignored webhook event: *github.PushEvent
Feb 12 14:58:52 www1 gopherci[1205]: 2017/02/12 14:58:52 github: pull request event: opened, installation id: 6055
Feb 12 14:58:52 www1 gopherci[1205]: 2017/02/12 14:58:52 GCPPubSubQueue: published a message with a message ID: 54710047742109
```

**Comment by bradleyfalzon** [comment]
This hasn't occurred in a while, and I'm going to need to refactor in #74, so closing this off for the moment.
